OSDOCS-3788:CSI inline ephemeral vols  TP-> GA (#47603)
openshift/openshift-docs: modules/ephemeral-storage-csi-inline-overview-admin-plugin.adoc

// Module included in the following assemblies:
//
// * storage/container_storage_interface/ephemeral-storage-csi-inline.adoc

:_content-type: CONCEPT
[id="ephemeral-storage-csi-overview-admin-plugin_{context}"]
= CSI Volume Admission plug-in

The CSI Volume Admission plug-in provides a mechanism where the use of an individual CSI driver capable of provisioning CSI ephemeral volumes can be restricted on pod admission. Administrators or distributions can optionally add a `csi-ephemeral-volume-profile` label to a `CSIDriver` object, and the label is then inspected by the Admission plug-in and used in enforcement, warning, and audit decisions.

> == Overview

Reviewer: Headings need an ID or a `[discrete]` tag per [OpenShift doc guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings). Maybe also include what it is an overview of.

> To make use of the CSI Volume Admission plug-in, administrators or distributions can add the `security.openshift.io/csi-ephemeral-volume-profile` label, which declares the CSI driver’s effective pod security profile when it is used to provide CSI ephemeral volumes. This “effective profile” communicates that a pod can use the CSI driver to mount CSI ephemeral volumes when the pod’s namespace is governed by a pod security standard:

Reviewer: ```suggestion
To use the CSI Volume Admission plug-in, administrators or distributions can add the `security.openshift.io/csi-ephemeral-volume-profile` label, which declares the CSI driver’s effective pod security profile when it is used to provide CSI ephemeral volumes. This “effective profile” communicates that a pod can use the CSI driver to mount CSI ephemeral volumes when the pod’s namespace is governed by a pod security standard, as shown in the following example:
```

[source, yaml]
----
kind: CSIDriver
metadata:
  name: csi.mydriver.company.org
  labels:
    security.openshift.io/csi-ephemeral-volume-profile: restricted
----

> The CSI Volume Admission plug-in inspects pod volumes when pods are created (existing pods that use CSI volumes are not affected). If a pod uses a container storage interface (CSI) volume, the plug-in looks up the `CSIDriver` object and inspects the `csi-ephemeral-volume-profile` label, and then use the label’s value in its xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-enforcement[enforcement], xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-warning[warning], and xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-audit[audit] decisions. 

Reviewer: IBM Style recommends [avoiding parentheses](https://www.ibm.com/docs/en/ibm-style?topic=punctuation-parentheses) so maybe this can be slightly rewritten.

> The CSI Volume Admission plug-in inspects pod volumes when pods are created (existing pods that use CSI volumes are not affected). If a pod uses a container storage interface (CSI) volume, the plug-in looks up the `CSIDriver` object and inspects the `csi-ephemeral-volume-profile` label, and then use the label’s value in its xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-enforcement[enforcement], xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-warning[warning], and xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-audit[audit] decisions. 

Reviewer: ```suggestion
The CSI Volume Admission plug-in inspects pod volumes when pods are created (existing pods that use CSI volumes are not affected). If a pod uses a container storage interface (CSI) volume, the plug-in looks up the `CSIDriver` object and inspects the `csi-ephemeral-volume-profile` label, and then uses the label’s value in its xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-enforcement[enforcement], xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-warning[warning], and xref:../../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc#security-profile-audit[audit] decisions. 
```

Reviewer: [xrefs are not allowed in modules](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#links-hyperlinks-and-cross-references) - I think you need to refer readers to the "Additional resources" section here for more info on these three things, and then list those links in the assembly "Additional resources" section.

[id="security-profile-enforcement"]
== Pod security profile enforcement

> When a CSI driver has the `csi-ephemeral-volume-profile label`, pods using the CSI driver to mount CSI ephemeral volumes must run in a namespace that enforces a pod security standard of equal or greater permission. If the namespace enforces a more restrictive standard, the CSI Volume Admission plug-in should deny admission.

Reviewer: Maybe add a sentence here about what the table is showing?

.Pod security profile enforcement
[cols=",^v,^v,^v,^v width="100%",options="header"]
|===
|PodSecurity Profile|driver label: restricted| driver label: baseline | driver label: privileged

|restricted
|Allowed
|Denied
|Denied

|baseline
|Allowed
|Allowed
|Denied

|privileged
|Allowed
|Allowed
|Allowed
|===

[id="security-profile-warning"]
== Pod security profile warning

> The CSI Volume Admission plug-in can also provide user-facing warnings (using appropriate HTTP response headers) if the CSI driver’s effective profile is more permissive than the pod security warning profile for the pod namespace.

Reviewer: ^ to avoid parentheses

|restricted
|No warning
|Warning
|Warning

|baseline
|No warning
|No warning
|Warning

|privileged
|No warning
|No warning
|No warning
|===

[id="security-profile-audit"]
== Pod security profile audit
The CSI Volume Admission plug-in can also apply audit annotations to the pod if the CSI driver’s effective profile is more permissive than the pod security audit profile for the pod namespace.

.Pod security profile audit
[cols=",^v,^v,^v,^v width="100%",options="header"]
|===
|PodSecurity Profile|driver label: restricted| driver label: baseline | driver label: privileged

|restricted
|No audit
|Audit
|Audit

|baseline
|No audit
|No audit
|Audit

|privileged
|No audit
|No audit
|No audit
|===

== Default behavior

If the referenced CSI driver for a CSI ephemeral volume does not have the `csi-ephemeral-volume-profile` label, the CSI Volume Admission plug-in considers the driver to have the privileged profile for enforcement, warning, and audit behaviors. Likewise, if the pod’s namespace does not have the pod security admission label set, the Admission plug-in assumes the restricted profile is allowed for enforcement, warning, and audit decisions.

> This means that if no labels are set, CSI ephemeral volumes using that CSI driver are only usable in privileged namespaces by default.

Reviewer: Does this sentence maybe belong with the previous sentence about labels not being set?

//
// * storage/container_storage_interface/ephemeral-storage-csi-inline.adoc

:_content-type: CONCEPT
[id="ephemeral-storage-csi-overview-admin-plugin_{context}"]
= CSI Volume Admission plugin

The Container Storage Interface (CSI) Volume Admission plugin allows you to restrict the use of an individual CSI driver capable of provisioning CSI ephemeral volumes on pod admission. Administrators can add a `csi-ephemeral-volume-profile` label, and this label is then inspected by the Admission plugin and used in enforcement, warning, and audit decisions.

[id="overview-admission-plugin"]
== Overview

To use the CSI Volume Admission plugin, administrators add the `security.openshift.io/csi-ephemeral-volume-profile` label to a `CSIDriver` object, which declares the CSI driver’s effective pod security profile when it is used to provide CSI ephemeral volumes, as shown in the following example:

[source, yaml]
----
kind: CSIDriver
metadata:
  name: csi.mydriver.company.org
  labels:
    security.openshift.io/csi-ephemeral-volume-profile: restricted <1>
----
<1> CSI driver object YAML file with the `csi-ephemeral-volume-profile` label set to "restricted"

> This “effective profile” communicates that a pod can use the CSI driver to mount CSI ephemeral volumes when the pod’s namespace is governed by a pod security standard.

Reviewer 2: Maybe we could use this one.
https://kubernetes.io/docs/concepts/security/pod-security-standards/

.Pod security profile warning
[cols=",^v,^v,^v,^v width="100%",options="header"]
|===
|Pod security profile|Driver label: restricted| Driver label: baseline | Driver label: privileged

|Restricted
|No warning
|Warning
|Warning

|Baseline
|No warning
|No warning
|Warning

|Privileged
|No warning
|No warning
|No warning
|===

[id="security-profile-audit"]
== Pod security profile audit
The CSI Volume Admission plugin can apply audit annotations to the pod if the CSI driver’s effective profile is more permissive than the pod security audit profile for the pod namespace. The following table shows the audit annotation applied for different pod security profiles for given label values.

.Pod security profile audit
[cols=",^v,^v,^v,^v width="100%",options="header"]
|===
|Pod security profile|Driver label: restricted| Driver label: baseline | Driver label: privileged

|Restricted
|No audit
|Audit
|Audit

|Baseline
|No audit
|No audit
|Audit

|Privileged
|No audit
|No audit
|No audit
|===

[id="admission-plugin-default-behavior"]
== Default behavior for the CSI Volume Admission plugin

> If the referenced CSI driver for a CSI ephemeral volume does not have the `csi-ephemeral-volume-profile` label, the CSI Volume Admission plugin considers the driver to have the privileged profile for enforcement, warning, and audit behaviors. Likewise, if the pod’s namespace does not have the pod security admission label set, the Admission plugin assumes the restricted profile is allowed for enforcement, warning, and audit decisions. Therefore, if no labels are set, CSI ephemeral volumes using that CSI driver are only usable in privileged namespaces by default.

Reviewer 3: It might be worth mentioning that CSI drivers that ship with OCP and support ephemeral volumes will have a reasonable default set for the csi-ephemeral-volume-profile label. The Shared Resource CSI Driver for example has this set to "restricted" by the operator because we consider it safe to use from restricted namespaces by default. The admin can still change the value of that label if they wish.

Reviewer 4: Should we add a new in-line ephemeral column to the CSI driver table?
https://docs.openshift.com/container-platform/4.13/storage/container_storage_interface/persistent-storage-csi.html#csi-drivers-supported_persistent-storage-csi

Author: Done
